set-balance.md

```diff
@@ -168,6 +192,346 @@ A `Result` is considered an `Action`.
+* `repatriate_reserved`
+
+```k
+    syntax Action ::= "slash_reserved" "(" AccountId "," Int ")"
+ // ------------------------------------------------------------
+    rule [slash-reserved]:
+         <k> slash_reserved(ACCOUNT, AMOUNT)
+          => set_reserved_balance(ACCOUNT, maxInt(0, RESERVED_BALANCE -Int AMOUNT))
+         ...
+         </k>
+         <accounts>
+           <account>
+             <accountID> ACCOUNT </accountID>
+             <reservedBalance> RESERVED_BALANCE </reservedBalance>
+             ...
+           </account>
+         </accounts>
+         <totalIssuance> TOTAL_ISSUANCE => TOTAL_ISSUANCE -Int minInt(RESERVED_BALANCE, AMOUNT) </totalIssuance>
+
+    syntax Action ::= "repatriate_reserved" "(" AccountId "," AccountId "," Int ")"
+ // -------------------------------------------------------------------------------
+    rule [repatriate-reserved]:
+         <k> repatriate_reserved(SLASHED, BENEFICIARY, AMOUNT)
+          => set_free_balance(BENEFICIARY, BENEFICIARY_FREE_BALANCE +Int minInt(SLASHED_RESERVED_BALANCE, AMOUNT))
+          ~> set_reserved_balance(SLASHED, SLASHED_RESERVED_BALANCE -Int minInt(SLASHED_RESERVED_BALANCE, AMOUNT))
+         ...
+         </k>
+         <accounts>
+           <account>
+             <accountID> SLASHED </accountID>
+             <reservedBalance> SLASHED_RESERVED_BALANCE </reservedBalance>
+             ...
+           </account>
+           <account>
+             <accountID> BENEFICIARY </accountID>
+             <reservedBalance> BENEFICIARY_RESERVED_BALANCE </reservedBalance>
+             <freeBalance> BENEFICIARY_FREE_BALANCE </freeBalance>
+             ...
+           </account>
+         </accounts>
+      requires BENEFICIARY_FREE_BALANCE +Int BENEFICIARY_RESERVED_BALANCE >Int 0
+```
+
+### Slashing
+
+Used to punish a node for violating the protocol.
+
+```k
+    syntax Action ::= "slash" "(" AccountId "," Int ")"
+ // ---------------------------------------------------
+    rule [slash]:
+         <k> slash(ACCOUNT, AMOUNT) => set_free_balance(ACCOUNT, FREE_BALANCE -Int AMOUNT) ... </k>
+         <accounts>
+           <account>
+             <accountID> ACCOUNT </accountID>
+             <freeBalance> FREE_BALANCE </freeBalance>
+             ...
+           </account>
+         </accounts>
+         <totalIssuance> TOTAL_ISSUANCE => TOTAL_ISSUANCE -Int AMOUNT </totalIssuance>
+      requires FREE_BALANCE >=Int AMOUNT
+
+    rule [slash-empty-free]:
+         <k> slash(ACCOUNT, AMOUNT)
+          => set_free_balance(ACCOUNT, 0)
+          ~> slash_reserved(ACCOUNT, AMOUNT -Int FREE_BALANCE)
+         ...
+         </k>
+         <accounts>
+           <account>
+             <accountID> ACCOUNT </accountID>
+             <freeBalance> FREE_BALANCE </freeBalance>
+             ...
+           </account>
+         </accounts>
+         <totalIssuance> TOTAL_ISSUANCE => TOTAL_ISSUANCE -Int FREE_BALANCE </totalIssuance>
+      requires FREE_BALANCE <Int AMOUNT
```

Comment:
Should there be checks on the `EXISTENTIAL_DEPOSIT` here, or is that handled at a higher level?

Reply:
NVM, `set*balance` will handle that.

Reply:
The Rust code does not check this, and I believe it is correct to do so, but I am checking with the Substrate team.